src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -3239,7 +3239,16 @@ public static BigInteger PopCount(BigInteger value)
         public static BigInteger RotateLeft(BigInteger value, int rotateAmount)
         {
             value.AssertValid();
-            int byteCount = (value._bits is null) ? sizeof(int) : (value._bits.Length * 4);
+
+            bool negx = value._sign < 0;
+            ReadOnlySpan<uint> bits = value._bits ?? stackalloc uint[1] { NumericsHelpers.Abs(value._sign) };
+            int xl = bits.Length;
+
+            if (negx && bits[^1] >= kuMaskHighBit
+                && !(bits.IndexOfAnyExcept(0u) == bits.Length - 1 && bits[^1] == kuMaskHighBit))
+                ++xl;
```

Comment:
nit: This needs braces and is a bit easier to understand written with the negation in the conditions:
```suggestion
            if (negx && (bits[^1] >= kuMaskHighBit) && (bits.IndexOfAnyExcept(0u) != (bits.Length - 1) || (bits[^1] != kuMaskHighBit)))
            {
                ++xl;
            }
```

A comment explaining what we're checking for here would also be helpful. In particular `bits` is stored as the absolute positive value. So we're looking for something that fits in 32-bits unsigned but needs more than 32-bits for two's complement signed. This would for example apply to `-2,147,483,649` (`0x8000_0001` unsigned and `0xF_7FFF_FFFF` signed) but not specifically to `-2,147,483,648` (`0x8000_0000` for both unsigned and signed).

Giving examples of where it does vs doesn't apply for each case can help with determining correctness/intent.